### DISCUSSION TITLE
NormalizerMySQLUnicodeCIExceptKanaCIKanaWithVoicedSoundMarkとTokenMecab("use_reading", true)を併用時の挙動について

### DISCUSSION THREAD

**Discussion by cmt-yamazaki** [pgroonga/pgroonga#345] [Q&A] [answered]
Ubuntu 22.04.2 LTS
postgresql-14-pgroonga        3.0.6-1.ubuntu22.04.1
groonga-normalizer-mysql     1.2.1-1.ubuntu22.04.1                   
groonga-token-filter-stem      13.0.1-1.ubuntu22.04.1 
groonga-tokenizer-mecab      13.0.1-1.ubuntu22.04.1        

Pgroongaで`NormalizerMySQLUnicodeCIExceptKanaCIKanaWithVoicedSoundMark`と`TokenMecab("use_reading", true)`を使ってインデックスを作っています。
mecabの辞書は標準インストールのnaist-jdicやunidicで試しました。

```
DROP INDEX if exists pgroonga_content_index;
DROP TABLE if exists memos;

CREATE TABLE memos(id int, content text);
CREATE INDEX pgroonga_content_index ON memos USING pgroonga(content) 
    WITH (normalizers = 'NormalizerMySQLUnicodeCIExceptKanaCIKanaWithVoicedSoundMark', 
          tokenizer = 'TokenMecab("use_reading", true)');

INSERT INTO memos (content) values ('東京都');
INSERT INTO memos (content) values ('大阪府');
INSERT INTO memos (content) values ('京都府');
INSERT INTO memos (content) values ('長野県');
INSERT INTO memos (content) values ('大城さん、大山さん、笹塚さん、坂本さん、井原さん');

select * from memos where content &@ 'とうきょう';   -- NG
select * from memos where content &@ 'トウキョウ';   -- NG
select * from memos where content &@ '東京';         -- OK
select * from memos where content &@ 'きょうと';     -- OK
select * from memos where content &@ 'キョウト';     -- OK
select * from memos where content &@ '京都';         -- OK

select * from memos where content &@ 'おおしろ';   -- NG
select * from memos where content &@ 'おおやま';   -- OK
select * from memos where content &@ 'ささづか';   -- NG
select * from memos where content &@ 'さかもと';   -- NG
select * from memos where content &@ 'いはら';     -- NG
```

都道府県名を登録し、SELECTすると、「京都」が「キョウト」や「きょうと」で引けません。
※何度もやっていると「キョウト」や「きょうと」がOKで、「トウキョウ」や「とうきょう」がNGになったりします

同様に「大城さん」、「大山さん」、「笹塚さん」、「坂本さん」、「井原さん」等の姓を1レコードに登録すると
「おおやま」だけOKで、「おおしろ」、「ささづか」、「さかもと」、「いはら」がNGになります。
最初の内は、「井原さん」をいれないでやっていて、「さかもと」だけOKでしたので、
ダブり文字がNGなのかと思ったのですが、姓を増やしたところ1つしかインデックス登録されないような挙動となりました。

尚、`NormalizerMySQLUnicodeCIExceptKanaCIKanaWithVoicedSoundMark`を無効にすれば、
`tokenizer = 'TokenMecab("use_reading", true)')`でトークナイズされたカタカナの読みでの検索は可能です。

```
DROP INDEX if exists pgroonga_content_index;
DROP TABLE if exists memos;

CREATE TABLE memos(id int, content text);
CREATE INDEX pgroonga_content_index ON memos USING pgroonga(content) 
    WITH (tokenizer = 'TokenMecab("use_reading", true)');

INSERT INTO memos (content) values ('東京都');
INSERT INTO memos (content) values ('大阪府');
INSERT INTO memos (content) values ('京都府');
INSERT INTO memos (content) values ('長野県');
INSERT INTO memos (content) values ('大城さん、大山さん、笹塚さん、坂本さん、井原さん');

select * from memos where content &@ 'トウキョウ';   -- OK
select * from memos where content &@ 'キョウト';      -- OK
select * from memos where content &@ 'オオシロ';      -- OK
select * from memos where content &@ 'オオヤマ';      -- OK
select * from memos where content &@ 'ササヅカ';      -- OK
select * from memos where content &@ 'サカモト';      -- OK
select * from memos where content &@ 'イハラ';         -- OK
```

理想は最初方に提示したSQLのすべてのSELECTについてレコードが返ってくることなのですが、良い方法はありますでしょうか。

**Comment by komainu8** [comment]
PGroongaのインデックスはちゃんと使われているので、PGroongaに問題がありそう。

```
pgroonga_test=# EXPLAIn ANALYZE VERBOSE select * from memos where content &@ 'いはら';
                                                              QUERY PLAN                                                              
--------------------------------------------------------------------------------------------------------------------------------------
 Index Scan using pgroonga_content_index on public.memos  (cost=0.00..4.01 rows=1 width=36) (actual time=0.159..0.159 rows=0 loops=1)
   Output: id, content
   Index Cond: (memos.content &@ 'いはら'::text)
 Planning Time: 0.487 ms
 Execution Time: 0.190 ms
(5 rows)

pgroonga_test=# select * from memos where content &@~ 'いはら';
 id | content 
----+---------
(0 rows)
```

直接、GroongaでTokenMecabの`use_reading`を使ってみる。

TokenMecabの`use_reading`オプションが正常に動いてなさそう。よみがなが出力されない。
よみがなが出力されてないので、おそらくインデックスにもよみがなが登録されていないため、"井原さん"が"イハラ"でヒットしない。と考えられる。

```
% groonga
> status
[[0,1686303413.061051,0.0002298355102539062],{"alloc_count":408,"starttime":1686303411,"start_time":1686303411,"uptime":2,"version":"13.0.1","n_queries":0,"cache_hit_rate":0.0,"command_version":1,"default_command_version":1,"max_command_version":3,"n_jobs":0,"features":{"nfkc":true,"mecab":true,"message_pack":true,"mruby":true,"onigmo":true,"zlib":true,"lz4":true,"zstandard":true,"kqueue":false,"epoll":true,"poll":false,"rapidjson":true,"apache_arrow":true,"xxhash":false,"back_trace":true,"reference_count":false},"apache_arrow":{"version_major":11,"version_minor":0,"version_patch":0,"version":"11.0.0"},"memory_map_size":86876160}]

> tokenize 'TokenMecab("use_reading", true)' '彼の名前は山田さんのはずです。'
[[0,1686303426.733231,0.0059967041015625],[{"value":"彼","position":0,"force_prefix":false,"force_prefix_search":false},{"value":"の","position":1,"force_prefix":false,"force_prefix_search":false},{"value":"名前","position":2,"force_prefix":false,"force_prefix_search":false},{"value":"は","position":3,"force_prefix":false,"force_prefix_search":false},{"value":"山田","position":4,"force_prefix":false,"force_prefix_search":false},{"value":"さん","position":5,"force_prefix":false,"force_prefix_search":false},{"value":"の","position":6,"force_prefix":false,"force_prefix_search":false},{"value":"はずです","position":7,"force_prefix":false,"force_prefix_search":false},{"value":"。","position":8,"force_prefix":false,"force_prefix_search":false}]]
```

Groongaのバグのような気がするのでGroongaを調査する。

**Comment by kou** [comment]
これ、漢字の検索対象を漢字でもひらがなでもカタカナでも検索したい、というのがやりたいことなんですよね？

`TokenMecab("use_reading", true)`はそのユースケースには向いていないと思います。

普通、MeCab（の広く使われている辞書）で「大城」と「おおしろ」を処理したときは、同じところで区切れないんですよ。

```console
$ echo 大城 | mecab
大城	名詞,固有名詞,人名,姓,*,*,大城,オオシロ,オーシロ,,
EOS
$ echo おおしろ | mecab
おお	感動詞,*,*,*,*,*,おお,オオ,オー,,
しろ	形容詞,自立,*,*,形容詞・アウオ段,ガル接続,しろい,シロ,シロ,しろ/白,
EOS
```

`TokenMecab`とか形態素解析ベースのトークナイザーは同じところで区切れないとヒットしません。

MeCabでよみがなに正規化して、`TokenNgram`でトークナイズするのがいいと思いますが、今はそういう機能はありません。`NormalizerTokenize("tokenizer", "TokenMecab('use_reading', true)")`みたいなノーマライザーを作ってトークナイズ結果を連結したものを正規化結果にするといいのかも。

いま使える機能で頑張るなら式インデックスを使えそうな気がしますけど、↓の`jsonb`を`text`にする書き方がわからない。。。（式インデックスでは`string_agg()`を使えない。。。）

```sql
select jsonb_path_query_array(jsonb_agg(pgroonga_tokenize('大城さん、大山さん、笹塚さん、坂本さん、井原さん', 'tokenizer', 'TokenMecab("use_reading", true)')), '$[*].value');
```

`NormalizerMySQLUnicodeCIExceptKanaCIKanaWithVoicedSoundMark`を使っているのはひらがな・カタカナを同一視したいからですよね？であれば、`NormalizerNFKC150("unify_kana", true)`でいいです。

`tokenize 'TokenMecab("use_reading", true)' '彼の名前は山田さんのはずです。'`がうまく動いていないのは単にnaist-jdic/ipadic以外の辞書を使っているだけな気がします。辞書によってメタデータの入り方が違ってよみがなに関しては汎用的な処理はできないので、今はnaist-jdic/ipadic互換の出力だけ対応しています。うーん、でも、メタデータの数で辞書を自動判定していい感じに扱えるようにできるかも？

**Comment by cmt-yamazaki** [comment, marked as answer]
@komainu8 さん、 @kou さん 
ありがとうございます。

kouさんの仰る通り、"漢字の検索対象を漢字でもひらがなでもカタカナでも検索したい"ということです。

> 普通、MeCab（の広く使われている辞書）で「大城」と「おおしろ」を処理したときは、同じところで区切れないんですよ。

「大城さん」と「オオシロさん」を登録して検索したら、「オオシロさん」しか引っかかりませんでした...

サンプルまで書いて頂いたので、式インデックスを試してみます。
幸い検索対象のテーブルは書き込みがそこまで忙しくないので、
式インデックスがダメでも、トリガーで検索用フィールドに読み仮名を入れる方法を検討したいと思います。

ちょっと特殊な環境で、外字とIPAフォントの異字体を使っています。
将来的な思惑として、外字の読み仮名をユーザー辞書で作っておいて、外字が入っていない環境でも検索出来たらいいなぁというのがありました。
（WEBアプリで表示にはWEBフォントで外字対応しているのですが、入力は外字がインストールされていないと難しいので）

**Comment by kou** [comment]
"漢字の検索対象を漢字でもひらがなでもカタカナでも検索したい"というユースケースはよくあると思うのでissueにしておきました。
https://github.com/groonga/groonga/issues/1577

MeCabの辞書をメンテナンスするのは大変そうな気はしますが、たしかに、そのアプローチでも動きそうな気はします。

あるいは、`NormalizerTable`で外字をすべて一般的な漢字に（文字単位で）正規化してからMeCabで処理するとかという方法もあるとはあるんですが、PGroongaから`NormalizerTable`を使うのはちょっとクセがあるんですよねぇ。。。

ところで、今でも外字は完全にシステム依存になっているんですか？
もし、広く（？）使われている外字とコードポイントのマッピングがあるなら`NormalizerNFKC`にオプションを追加して対応してしまうのもありかもしれません。

**Comment by kou** [comment]
https://ja.wikipedia.org/wiki/%E5%A4%96%E5%AD%97#%E8%91%97%E5%90%8D%E3%81%AA%E5%A4%96%E5%AD%97%E9%9B%86%E5%90%88
「著名な外字集合」というのがあるのか！

**Comment by cmt-yamazaki** [comment]
issue、ありがとうございます。
NormalizerTableは試してませんでしたので、また確認してみます。

外字はWindowsの場合はシステム依存です。
Windowsの外字エディタで登録した外字は「EUDC.TTE」、「EUDC.EUF」というファイルに保存され、エクスプローラーでFontsディレクトリを見たときには見えなくなっています。通常のフォントファイルとは扱いが違いFontファミリーとして扱われません。

issueも上げて頂きましたので、本件についてはやりようが見つかったということで解決にします。
ありがとうございました。
